README.md
# gitleaves

*Git Based Leaves management system*

Project status: WIP

## Specifications

### File names

- leaves.\<YYYY\>.csv
- extras.\<YYYY\>.csv

### Format

### Leaves
MMDD [- MMDD], Name, Reason / Details [Optional]

### Extra days
MMDD [- MMDD], Name, Reason / Details [Optional]

### Example


```
1115 - 1126, Tony Stark, Hawaii
1129 - 1210, Peter Parker,
1129 - 1202, Bruce Wayne,
1220 - 1224, Loki Laufeyson, Smell the flowers
```

## Usage

```
$ pip install gitleaves
$ mkdir data/
$ cp leaves.<this-year>.csv data/
$ gitleaves genreports
$ ls reports/
```

## Process

- Apply for leave(s) by adding records to leaves.YYYY.csv
- Optionally mention details/reasons in commit message
- Raise a PR
- Leave is approved if the PR is merged

## Guidelines

- Do take leaves when necessary :)
- Make sure you spell your name same everywhere
- Generally try to add records in order
- When applying for leaves
    - make sure you take look at peers leave plans 
    - ensure that you have a collegue to cover your work while you are away
    - if necessary, do essential knowledge transfer to ensure that there are no dependencies on you
    - if you are a major contributor to an important release then avoid immediate leaves after the release

## Dev Guidelines
- Human Readable data files (as much possible)

## Features

## TODO

## Credits

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage` project template.

- Cookiecutter: https://github.com/audreyr/cookiecutter
- audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


CHANGELOG.md
## History

### 0.1.0 (2021-11-10)

* First release on PyPI.


tests/__init__.py
"""Unit test package for gitleaves."""


tests/test_gitleaves.py
#!/usr/bin/env python

"""Tests for `gitleaves` package."""

import pytest

from click.testing import CliRunner

from gitleaves import gitleaves
from gitleaves import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    result = runner.invoke(cli.main)
    assert result.exit_code == 0
    assert 'gitleaves.cli.main' in result.output
    help_result = runner.invoke(cli.main, ['--help'])
    assert help_result.exit_code == 0
    assert '--help  Show this message and exit.' in help_result.output


gitleaves/reports.py
import arrow
import calendar
import csv
import datetime
import os
import os.path
import pathlib


from collections import defaultdict

from jinja2 import FileSystemLoader, Environment

srcdir = pathlib.Path(__file__).parent.resolve()
reports_dir = 'reports'
templates_dir = os.path.join(srcdir, 'templates')
ghwiki_reports_dir = os.path.join(reports_dir, 'ghwiki')
templates = Environment(loader=FileSystemLoader(templates_dir),
                        trim_blocks=True)

today = datetime.date.today()
leaves_csv_path = f'leaves.{today.year}.csv'
extras_csv_path = f'leaves.{today.year}.csv'


def ddmm2date(s):
    return arrow.get(f'{today.year}{s.strip()}')


def range2dates(daterange):
    """
    '1307 - 2907' →  [date(2021, 7, 13), ....]
    Except weekends
    """
    if '-' in daterange:
        start, end = [ddmm2date(d) for d in daterange.split('-')]
        dates = arrow.Arrow.range('day', start, end)
    else:
        dates = [ddmm2date(daterange)]
    return [d.date() for d in dates]


def load_csv(csv_path):
    bydates = defaultdict(list)
    bynames = defaultdict(list)
    bymonths = defaultdict(lambda: defaultdict(list))
    csv_f = csv.reader(open(csv_path))
    next(csv_f)
    for row in csv_f:
        daterange, applicant, *_ = row
        applicant = applicant.split('#')[0].strip()
        dates = range2dates(daterange)
        for date in dates:
            bydates[date].append(applicant)
            bynames[applicant].append(date)
            bymonths[date.month][date].append(applicant)

    return {'bydates': bydates, 'bynames': bynames, 'bymonths': bymonths}


def gen_ghwiki_reports():
    data = load_csv(os.path.join('data', leaves_csv_path))
    today_leaves = data['bydates'][datetime.date.today()]
    this_month = datetime.date.today().month
    next_leaves_by_month = ((calendar.month_name[month], leaves)
                            for month, leaves in data['bymonths'].items()
                            if month >= this_month)
    template = templates.get_template('ghwiki/Home.md')
    if not os.path.exists(ghwiki_reports_dir):
        os.makedirs(ghwiki_reports_dir)
    with open(f'{ghwiki_reports_dir}/Home.md', 'w') as report:
        report.write(template.render(today_leaves=today_leaves,
                                     next_leaves_by_month=next_leaves_by_month))
    return ghwiki_reports_dir

def upload_ghwiki_reports():
    raise NotImplementedError


gitleaves/genreport.py
import arrow
import calendar
import csv
import datetime
import os
import pathlib

from jinja2 import FileSystemLoader, Environment

from collections import defaultdict

reports_dir = 'reports'
srcdir = pathlib.Path(__file__).parent.resolve()
templates_dir = os.path.join(srcdir, 'template')
ghwiki_reports_dir = os.path.join(reports_dir, 'ghwiki')
templates = Environment(loader=FileSystemLoader(templates_dir),
                        trim_blocks=True)

today = datetime.date.today()
leaves_csv_path = f'leaves.{today.year}.csv'
extras_csv_path = f'leaves.{today.year}.csv'


def ddmm2date(s):
    return arrow.get(f'{today.year}{s.strip()}')


def range2dates(daterange):
    """
    '1307 - 2907' →  [date(2021, 7, 13), ....]
    Except weekends
    """
    if '-' in daterange:
        start, end = [ddmm2date(d) for d in daterange.split('-')]
        dates = arrow.Arrow.range('day', start, end)
    else:
        dates = [ddmm2date(daterange)]
    return [d.date() for d in dates]


def load_csv(csv_path):
    bydates = defaultdict(list)
    bynames = defaultdict(list)
    bymonths = defaultdict(lambda: defaultdict(list))
    csv_f = csv.reader(open(csv_path))
    next(csv_f)
    for row in csv_f:
        daterange, applicant, *_ = row
        applicant = applicant.split('#')[0].strip()
        dates = range2dates(daterange)
        for date in dates:
            bydates[date].append(applicant)
            bynames[applicant].append(date)
            bymonths[date.month][date].append(applicant)

    return {'bydates': bydates, 'bynames': bynames, 'bymonths': bymonths}


def gen_ghwiki_reports():
    data = load_csv(leaves_csv_path)
    today_leaves = data['bydates'][datetime.date.today()]
    this_month = datetime.date.today().month
    next_leaves_by_month = ((calendar.month_name[month], leaves)
                            for month, leaves in data['bymonths'].items()
                            if month >= this_month)
    template = templates.get_template('ghwiki/index.md')
    with open(f'{ghwiki_reports_dir}/index.md', 'w') as report:
        report.write(template.render(today_leaves=today_leaves,
                                     next_leaves_by_month=next_leaves_by_month))
